sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,296 @@ import org.apache.spark.util.Utils
+        case VALUE_STRING =>
+          val stringValue = parser.getText
+          if (stringValue.contains("-")) {
+            // The format of this string will probably be "yyyy-mm-dd".
+            DateTimeUtils.millisToDays(DateTimeUtils.stringToTime(parser.getText).getTime)
+          } else {
+            // In Spark 1.5.0, we store the data as number of days since epoch in string.
+            // So, we just convert it to Int.
+            stringValue.toInt
+          }
+      }
 
-      case (VALUE_TRUE, BooleanType) =>
-        true
+    case BinaryType =>
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case VALUE_STRING => parser.getBinaryValue
+      }
 
-      case (VALUE_FALSE, BooleanType) =>
-        false
+    case dt: DecimalType =>
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT) =>
+          Decimal(parser.getDecimalValue, dt.precision, dt.scale)
+      }
 
-      case (START_OBJECT, st: StructType) =>
-        convertObject(factory, parser, st)
+    case st: StructType =>
+      val fieldConverters = st.map(_.dataType).map(makeConverter)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertObject(parser, st, fieldConverters)
+      }
 
-      case (START_ARRAY, ArrayType(st, _)) =>
-        convertArray(factory, parser, st)
+    case at: ArrayType =>
+      val elementConverter = makeConverter(at.elementType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_ARRAY => convertArray(parser, elementConverter)
+      }
 
-      case (START_OBJECT, MapType(StringType, kt, _)) =>
-        convertMap(factory, parser, kt)
+    case mt: MapType =>
+      val valueConverter = makeConverter(mt.valueType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertMap(parser, valueConverter)
+      }
+
+    case udt: UserDefinedType[_] =>
+      makeConverter(udt.sqlType)
+
+    case _ =>
+      (parser: JsonParser) =>
+        // Here, we pass empty `PartialFunction` so that this case can be
+        // handled as a failed conversion. It will throw an exception as
+        // long as the value is not null.
+        parseJsonToken(parser, dataType)(PartialFunction.empty[JsonToken, Any])
+  }
+
+  /**
+   * This handles nulls ahead before trying to check the tokens, and applies the conversion
+   * function and then checks failed the conversion afterward if it `f` fails to convert the value.
+   *
+   * In more details, it checks `FIELD_NAME` if exists and then skip. If this is called after
+   * `START_OBJECT`, then, the next token can be `FIELD_NAME`. Since the names are kept in
+   * `JacksonParser.convertObject`, this `FIELD_NAME` token can be skipped as below. When this
+   * is called after `START_ARRAY`, the tokens become ones about values until `END_ARRAY`.
+   * In this case, we don't have to skip.
```

Comment:
Ah, the tokens after `START_ARRAY` will be consecutive tokens until `END_ARRAY`. For example,

```scala
parser.nextToken() // => START_ARRAY
parser.nextToken() // => VALUE_STRING
parser.nextToken() // => VALUE_STRING
parser.nextToken() // => VALUE_STRING
parser.nextToken() // => END_ARRAY
```

In this case, we don't (can't) skip `FIELD_NAME` token. I will try to make the documentation cleaner.